rs/replicated_state/src/canister_state/system_state/wasm_chunk_store.rs

```diff
@@ -135,15 +135,26 @@ impl WasmChunkStore {
         Ok(())
     }
 
-    pub fn insert_chunk(&mut self, max_size: NumBytes, chunk: &[u8]) -> Result<[u8; 32], String> {
-        let hash = ic_crypto_sha2::Sha256::hash(chunk);
+    pub fn contains_chunk(&self, hash: &[u8; 32]) -> bool {
+        self.metadata.chunks.contains_key(hash)
+    }
+
+    pub fn insert_chunk(&mut self, max_size: NumBytes, chunk: &[u8], hash: &[u8; 32]) {
+        debug_assert!(
+            *hash == ic_crypto_sha2::Sha256::hash(chunk),
+            "Provided hash {:?} does not match actual chunk hash {:?}",
+            hash,
+            ic_crypto_sha2::Sha256::hash(chunk)
+        );
 
         // No changes needed if we already have the chunk
-        if self.metadata.chunks.contains_key(&hash) {
-            return Ok(hash);
+        if self.contains_chunk(hash) {
+            // We should not attempt to insert an existing chunk: `Self::contains_chunk` should be invoked first.
+            debug_assert!(false, "Inserting an existing chunk {:?}", hash);
+            return;
         }
 
-        self.can_insert_chunk(max_size, chunk)?;
+        self.can_insert_chunk(max_size, chunk).unwrap();
```

Comment:
I generally find unwrapping based on non-local assumptions (someone already made the check up the call stack) quite brittle, for me the old way seems preferable. If you really want to avoid the cost of calling twice, a classic way around it is to encode this in the type (have a newtype `ValidatedChunk` or something that can only be created by `can_insert_chunk`). 

If there's really no other way but to unwrap, I'd at least use an `.expect` instead of just an unwrap.

Reply:
I agree with Oggy. This general rule of thumb is also mentioned in the [best practices](https://github.com/dfinity/ic/blob/master/rs/replicated_state/best-practices-panics.md#when-should-panics-be-avoided).

The suggestion of a `ValidatedChunk` is what I would also propose, similar to how we introduced `ValidatedMemoryUsage` in a previous change.

Reply:
Would you mind computing the chunk hash twice? If not, then the chunk hash could be recomputed and bundled in that `ValidatedChunk` struct.

Reply:
I refactored the code: `can_insert_chunk` produces `ChunkValidationResult` containing a `ValidatedChunk` (along with its hash), a variant that the chunk already exists (along with its hash), or a validation error.

Reply:
Thanks, I'm good from my side but since the thread was not initiated by me, I'll let Oggy resolve.